In [1]:
import inspect

In [45]:
def get_closure_function_info(func):
    print('Below is details of the func...')
    print('Function code: \n{}'.format(inspect.getsource(func)))
    print('Function documentation: {}'.format(func.__doc__))
    print("Function closure: {}".format(func.__closure__))
    print("Function closure type: {}".format(type(func.__closure__)))
    # Number of `nonlocal` variables or values the closure function uses or remembers
    print("Function closure length: {}".format(len(func.__closure__)))
    for i in range(len(func.__closure__)):
        print("Function closure cell[{0}] contents: {1}".format(i, func.__closure__[i].cell_contents))

In [46]:
# Closure example 01
def foo():
    a = 5
    def bar():
        return a
    return bar

func = foo()
print(func())
get_closure_function_info(func)

5
Below is details of the func...
Function code: 
    def bar():
        return a

Function documentation: None
Function closure: (<cell at 0x118a6a4d0: int object at 0x1027e5670>,)
Function closure type: <class 'tuple'>
Function closure length: 1
Function closure cell[0] contents: 5


In [17]:
# Closure example 01
def foo():
    a = 5
    b = 10
    def bar():
        return a, b
    return bar

func = foo()
print(func())
get_closure_function_info(func)

(5, 10)
    def bar():
        return a, b

Function closure: (<cell at 0x1062c5540: int object at 0x1027e5670>, <cell at 0x1184c16f0: int object at 0x1027e5710>)
Function closure type: <class 'tuple'>
Function closure length: 2
Function closure cell[0] contents: 5
Function closure cell[1] contents: 10


In [ ]:
# Closure example 02
x = 25
def foo():
    def bar():
        print(x)
    return bar

# del(x) --> It deletes the x variable and the function call func() = bar() can't find x anymore.
func = foo()
del(x)
func()

'''
NameError                                 Traceback (most recent call last)
Cell In[18], line 12
     10 func = foo()
     11 del(x)
---> 12 func()

Cell In[18], line 5, in foo.<locals>.bar()
      4 def bar():
----> 5     print(x)

NameError: name 'x' is not defined
'''

In [18]:
# Closure example 03
x = 25
def foo(value):
    def bar():
        print(value)
    return bar

func = foo(x)
del(x) 
# Even if the global variable is deleted, 
# its value is passed by copy to `value` variable in foo and 
# then `bar` function uses it and remembers it.
# so deleting `x` does not have any impact.
func()

25


In [19]:
# Closure example 04
x = 105
def foo(value):
    def bar():
        print(value)
    return bar
x = foo(x)
x()

105


In [47]:
# Closure example 05
def parent(arg1, arg2):
    value = 44
    my_dict = {'a':'Chocolate', 'b': 50}
    def child():
        my_local_variable = 100
        print(my_dict)
        print(arg1, arg2)
        print(value)
        print(my_local_variable)
    
    return child

func = parent(10, 100)
func()
print([cell.cell_contents for cell in func.__closure__])
get_closure_function_info(func)

{'a': 'Chocolate', 'b': 50}
10 100
44
100
[10, 100, {'a': 'Chocolate', 'b': 50}, 44]
Below is details of the func...
Function code: 
    def child():
        my_local_variable = 100
        print(my_dict)
        print(arg1, arg2)
        print(value)
        print(my_local_variable)

Function documentation: None
Function closure: (<cell at 0x118a6b550: int object at 0x1027e5710>, <cell at 0x118a6add0: int object at 0x1027e6250>, <cell at 0x1187c2da0: dict object at 0x11845b4c0>, <cell at 0x1187c2d70: int object at 0x1027e5b50>)
Function closure type: <class 'tuple'>
Function closure length: 4
Function closure cell[0] contents: 10
Function closure cell[1] contents: 100
Function closure cell[2] contents: {'a': 'Chocolate', 'b': 50}
Function closure cell[3] contents: 44


In [48]:
print(func.__closure__)
print(type(func.__closure__))
print(len(func.__closure__))
print(type(func.__closure__[0]))

(<cell at 0x118a6b550: int object at 0x1027e5710>, <cell at 0x118a6add0: int object at 0x1027e6250>, <cell at 0x1187c2da0: dict object at 0x11845b4c0>, <cell at 0x1187c2d70: int object at 0x1027e5b50>)
<class 'tuple'>
4
<class 'cell'>


In [58]:
# Decorator example 01
def multiply(a, b):
    '''Multiply 2 numbers'''
    return a * b

new_multiply = double_args(multiply)
print(new_multiply(2, 5))
get_closure_function_info(new_multiply)
print(inspect.getsource(new_multiply.__closure__[0].cell_contents))

40
Below is details of the func...
Function code: 
    def wrapper(a, b):
        '''This is the wrapper function of double_args'''
        #print(f'{func.__name__} is called with the arguments: {args}')
        return func(a * 2, b * 2)

Function documentation: This is the wrapper function of double_args
Function closure: (<cell at 0x106303c10: function object at 0x1187f7740>,)
Function closure type: <class 'tuple'>
Function closure length: 1
Function closure cell[0] contents: <function multiply at 0x1187f7740>
def multiply(a, b):
    '''Multiply 2 numbers'''
    return a * b



In [71]:
@double_args
def a_multiply(a, b):
    '''a multiply'''
    return a * b + 2

print('testing ... {}'.format(a_multiply(2, 5)))
print(a_multiply.__name__)
print(a_multiply.__doc__)

testing ... 42
wrapper
This is the wrapper function of double_args


In [70]:
def print_args(func):
    def wrapper(*args, **kwargs):
        print(args)
        print(f'{func.__name__!r} is called with arguments: {args}')
        return func(*args, **kwargs)
    return wrapper


b_multiply = print_args(multiply)
print('testing ... {}'.format(b_multiply(3, 6)))

# print(dir(a_multiply))


# print(b_multiply.__name__)
# print(b_multiply.__doc__)

sig = inspect.signature(multiply)
print(sig)

(3, 6)
'wrapper' is called with arguments: (3, 6)
{'a': 3, 'b': 6}
multiply was called with a=3, b=6
testing ... 18
(*args, **kwargs)


In [65]:
import inspect

def multiply(a, b):
    return a * b

def print_args(func):
    sig = inspect.signature(func)
    def wrapper(*args, **kwargs):
        bound = sig.bind(*args, **kwargs).arguments
        print(bound)
        str_args = ', '.join(['{}={}'.format(k, v) for k, v in bound.items()])
        print('{} was called with {}'.format(func.__name__, str_args))
        return func(*args, **kwargs)
    return wrapper
    
multiply = print_args(multiply)
print('testing ... {}'.format(multiply(b=3,a=6)))
# print(inspect.getsource(multiply))

{'a': 6, 'b': 3}
multiply was called with a=6, b=3
testing ... 18


In [152]:
import inspect

def my_function(param1, param2, param3="default"):
    pass

signature = inspect.signature(my_function)
print(signature)
parameters = signature.parameters
print(parameters)

a = [f'{name}={param.default}' for name, param in parameters.items()]
print(a)

(param1, param2, param3='default')
OrderedDict({'param1': <Parameter "param1">, 'param2': <Parameter "param2">, 'param3': <Parameter "param3='default'">})
["param1=<class 'inspect._empty'>", "param2=<class 'inspect._empty'>", 'param3=default']


In [153]:
import time

def timer(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        print(f'{func.__name__}: {end_time - start_time}')
        return result
    return wrapper
              
def memoize(func):
    cache = {}
    def wrapper(n):
        if n not in cache:
            cache[n] = func(n)
        return cache[n]
    return wrapper


def general_memoize(func):
    cache = {}
    def wrapper(*args, **kwargs):
        kwargs_key = tuple(sorted(kwargs.items()))
        if (args, kwargs_key) not in cache:
            cache[(args, kwargs_key)] = func(*args, **kwargs)
        return cache[(args, kwargs_key)]
    return wrapper


def counter(func):
    def wrapper(*args, **kwargs):
        wrapper.count += 1
        # Call the function being decorated and return the result
        return func(*args, **kwargs)
    # Set count to 0 to initialize call count for each new decorated function
    wrapper.count = 0
    # Return the new decorated function
    return wrapper
    
@counter
@memoize
def calculate(n):
    if n == 1:
        return 1
    elif n % 2 == 1:
        return calculate(3 * n + 1) + 1
    else:
        return calculate(n // 2) + 1

@timer
def run():
    result = [calculate(i) for i in range(1, 100001)]
    print(max(result))

run()

print('calculate is called {} times'.format(calculate.count))


# @general_memoize
# def slow_function(a, b):
#     print('Sleeping ...')
#     time.sleep(5)
#     return a + b

# print(slow_function(1, 2))
# print(slow_function(1, 2))


351
run: 0.13143181800842285
calculate is called 317211 times


In [157]:

print(calculate(1))
print(calculate(2))
print(calculate(3))
print(calculate(4))
print(calculate(8))


1
2
8
3
4


# Witout memoize function
351
run: 3.2158761024475098
calculate is called 10853840 times


# With memoize function
351
run: 0.13365602493286133
calculate is called 317211 times



In [ ]:
# Return return type
def print_return_type(func):
  # Define wrapper(), the decorated function
  def wrapper(*args, **kwargs):
    # Call the function being decorated
    result = func(*args, **kwargs)
    print('{}() returned type {}'.format(
      func.__name__, type(result)
    ))
    return result
  # Return the decorated function
  return wrapper
  
@print_return_type
def foo(value):
  return value
  
print(foo(42))
print(foo([1, 2, 3]))
print(foo({'a': 42}))

In [ ]:
# Counting function calls
from functools import wraps

def counter(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        wrapper.count += 1
        # Call the function being decorated and return the result
        return func(*args, **kwargs)
    
    # Set count to 0 to initialize call count for each new decorated function
    wrapper.count = 0
    # Return the new decorated function
    return wrapper

# Decorate foo() with the counter() decorator
@counter
def foo():
    ''' a foo function. '''
    print('calling foo()')
  
foo()
foo()

print('foo() was called {} times.'.format(foo.count))

In [ ]:
foo()
print('foo() was called {} times.'.format(foo.count))

In [ ]:
foo.__doc__

In [ ]:
foo.__wrapped__()

In [ ]:
foo()
print('foo() was called {} times.'.format(foo.count))

# Decorator Factory. Function that returns the decorator.

In [72]:
def run_n_times(n):
    def decorator(func):
        def wrapper(*args, **kwargs):
            for i in range(1, n+1):
                print(f'{i}', end=' ')
                func(*args, **kwargs)
        return wrapper
    return decorator

@run_n_times(3)
def print_sum(a, b):
    print(a + b)

print_sum(10, 100)

@run_n_times(6)
def print_hello():
    print('hello')

print_hello()

1 110
2 110
3 110
1 hello
2 hello
3 hello
4 hello
5 hello
6 hello


In [122]:
def html(open_tag, close_tag):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
          msg = func(*args, **kwargs)
          return '{}{}{}'.format(open_tag, msg, close_tag)
        # Return the decorated function
        return wrapper
    # Return the decorator
    return decorator

# Make hello() return bolded text
@html('<b>', '</b>')
def hello(name):
  return 'Hello {}!'.format(name)
  
print(hello('Alice'))

@html('<i>', '</i>')
def goodbye(name):
    return 'Goodbye {}'.format(name)

print(goodbye('Alice'))

# Wrap the result of hello_goodbye() in <div> and </div>
@html('<div>', '</div>')
def hello_goodbye(name):
  return '\n{}\n{}\n'.format(hello(name), goodbye(name))
  
print(hello_goodbye('Alice'))

<b>Hello Alice!</b>
<i>Goodbye Alice</i>
<div>
<b>Hello Alice!</b>
<i>Goodbye Alice</i>
</div>


# Timeout

In [ ]:
import time
import signal
import functools

def raise_timeout(*args, **kwargs):
    raise TimeoutError()

signal.signal(signalnum=signal.SIGALRM, handler=raise_timeout)
signal.alarm(5)

def timeout_in_5s(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        signal.alarm(5)
        try:
            return func(*args, **kwargs)
        finally:
            signal.alarm(0)
    return wrapper

@timeout_in_5s
def foo():
    time.sleep(5)
    print('foo')



In [127]:
import time
import functools
import signal

def raise_timeout(*args, **kwargs):
    raise TimeoutError()

signal.signal(signalnum=signal.SIGALRM, handler=raise_timeout)

def timeout(n_seconds):
    def decorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            signal.alarm(n_seconds)
            try:
                return func(*args, **kwargs)
            finally:
                signal.alarm(0)
        return wrapper
    return decorator

@timeout(5)
def foo():
    time.sleep(1)
    print('foo')

foo()

foo


In [128]:
# tagging a function
def tag(*tags):
    # Define a new decorator, named "decorator", to return
    def decorator(func):
        # Ensure the decorated function keeps its metadata
        @wraps(func)
        def wrapper(*args, **kwargs):
            # Call the function being decorated and return the result
            return func(*args, **kwargs)
        wrapper.tags = tags
        return wrapper
    # Return the new decorator
    return decorator

@tag('test', 'this is a tag')
def foo():
    pass

print(foo.tags)

('test', 'this is a tag')


In [137]:
# Test if the function is returning the correct data type

def returns(return_type):
  # Complete the returns() decorator
  def decorator(func):
    def wrapper(*args, **kwargs):
      result = func(*args, **kwargs)
      assert type(result) == return_type
      return result
    return wrapper
  return decorator
  
@returns(dict)
def foo(value):
  return value

try:
  print(foo([1,2,3]))
except AssertionError:
  print('foo() did not return a dict!')
  

foo() did not return a dict!


In [129]:
a = {1: 10, 2: 20}
type(a)

dict

In [136]:
a = [1, 10, 2, 20]
print(type(a))
print(*a)

<class 'list'>
1 10 2 20


In [131]:
type(a) == list

True

In [133]:
a = (1, 10, 2, 20)

In [135]:
print(*a)

1 10 2 20


In [79]:
from functools import wraps
import inspect

def my_decorator(func):
    """
    This is my decorator.
    """
    @wraps(func)
    def wrapper(*args, **kwargs):
        """
        This is the wrapper function.
        """
        print("Something is happening before the function is called.")
        result = func(*args, **kwargs)
        print("Something is happening after the function is called.")
        return result
    return wrapper

@my_decorator
def say_hello(name):
    """
    A simple function that says hello.
    """
    print(f"Hello, {name}!")

# Now, after decorating, the original function's metadata is preserved
print(f"Decorated function name: {say_hello.__name__}")
print(f"Decorated function documentation: {say_hello.__doc__}")

# You can still get information about the decorator and wrapper, but it's a bit more involved.
# You can access the wrapped function using the `__wrapped__` attribute
# (available since Python 3.4)
try:
    print(f"Original function name (via __wrapped__): {say_hello.__wrapped__.__name__}")
    print(f"Original function documentation (via __wrapped__): {say_hello.__wrapped__.__doc__}")
    print(say_hello.__wrapped__.__code__)
    print(inspect.getsource(my_decorator))
except AttributeError:
    print("The __wrapped__ attribute is not available in this Python version.")

Decorated function name: say_hello
Decorated function documentation: 
    A simple function that says hello.
    
Original function name (via __wrapped__): say_hello
Original function documentation (via __wrapped__): 
    A simple function that says hello.
    
<code object say_hello at 0x118f4fe10, file "/var/folders/8k/057rw2fx78s_znmhc67rr75c0000gn/T/ipykernel_59482/430254931.py", line 19>
def my_decorator(func):
    """
    This is my decorator.
    """
    @wraps(func)
    def wrapper(*args, **kwargs):
        """
        This is the wrapper function.
        """
        print("Something is happening before the function is called.")
        result = func(*args, **kwargs)
        print("Something is happening after the function is called.")
        return result
    return wrapper

